clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -1126,9 +1141,11 @@ public void onFailure(RuntimeException e) {
                                             // as the duration of the rebalance timeout. If we stop sending heartbeats,
                                             // however, then the session timeout may expire before we can rejoin.
                                             heartbeat.receiveHeartbeat();
+                                        } else if (e instanceof FencedInstanceIdException) {
+                                            log.error("Caught fenced group.instance.id {} error in heartbeat thread", groupInstanceId);
+                                            heartbeatThread.failed.set(e);
```

Comment:
Should we return after we fail the heartbeat thread? We do no want it to keep running I assume.

Reply:
We are in a if-else branch here, but I agree. In case someone adds logic after if-else block in the future.

Reply:
Oh, actually it's against code style, so just leave it.

Reply:
We do need a way to stop the heartbeat thread still, right? Perhaps we can invoke `disable()`?

Reply:
Oh, got it. Let's stop it through disable() then